File: src/components/rule.lua
local asserts = require("src.assertions")

-- Rules define templates for defined actions like code compile

Rule = {}
Rule.__index = Rule

function Rule.new(name, command, ...)
	asserts.isNonEmptyString(name, "Rule name")
	asserts.isNonEmptyString(command, "Command")
	for _, value in ipairs({ ... }) do
		asserts.isInstanceOf(value, Binding, "Additional rule arguments must be bindings.")
	end

	local rule = setmetatable({}, Rule)

	rule.name = name
	rule.command = command
	rule.bindings = {}

	for _, binding in ipairs({ ... }) do
		table.insert(rule.bindings, binding)
	end

	return rule
end

local allowedKeys = {
	"depfile",
	"deps",
	"msvc_deps_prefix",
	"description",
	"dyndep",
	"generator",
	"in",
	"in_newline",
	"out",
	"restat",
	"rspfile",
	"rspfile_content",
}

local utils = require("src.utils")
function Rule:toString()
	local rule = "rule " .. self.name .. "\n"
	rule = rule .. "    command = " .. self.command .. "\n"

	for _, binding in ipairs(self.bindings) do
		if utils.contains(allowedKeys, binding.key) then
			rule = rule .. "    " .. tostring(binding)
		end
	end

	return rule
end

function Rule.reference(self)
	return self.name
end

Rule.__tostring = Rule.toString

File: src/components/rule.cpp
#include "rule.hpp"
#include <expected>
#include <sstream>
#include <stdexcept>

namespace jonin_bt {
using namespace std;

static vector<string> allowed_refs = {
    "${depfile}",     "${deps}",       "${msvc_deps_prefix}",
    "${description}", "${dyndep}",     "${generator}",
    "${in}",          "${in_newline}", "${out}",
    "${restat}",      "${rspfile}",    "${rspfile_content}",
};

Rule::Rule(string const &name_, string const &command_,
           vector<Binding> const &bindings_)
    : name(name_), command(command_), bindings(bindings_) {
	if (name.empty())
		throw new invalid_argument("Rule name must not be empty");
	if (command.empty())
		throw new invalid_argument("Rule command must not be empty");
}

auto Rule::new_Rule(string const &name, string const &command)
    -> expected<Rule, string> {
	vector<Binding> no_bindings;
	return Rule::new_Rule(name, command, no_bindings);
}

auto Rule::new_Rule(string const &name, string const &command,
                    vector<Binding> const &bindings) -> expected<Rule, string> {
	try {
		return Rule(name, command, bindings);
	} catch (invalid_argument const &e) {
		return unexpected(string(e.what()));
	}
}

auto Rule::to_string() const -> string {
	ostringstream oss;

	oss << "rule " << name << "\n";
	oss << "    command = " << command << "\n";

	for (Binding const &binding : bindings)
		if (find(allowed_refs.begin(), allowed_refs.end(),
		         binding.to_referece()) != allowed_refs.end())
			oss << "    " << binding.to_string();
		else
			throw invalid_argument("Rule binding not allowed");

	return oss.str();
}

auto Rule::to_referece() const -> string { return name; }

} // namespace jonin_bt
